[HWKINVENT-157] Inventory sync

metlos opened the pull request:
Work towards inventory sync with feeds.
- [x] Merkle tree hash capable identity hash
- [x] Store identity hashes on feed-based entities
- [x] Cascade updates of hashable entities up to the topmost hashable parent
- [x] Sync API
- [x] REST API

metlos committed:
First, rather random changes to start supporting feed sync.
This is all most probably going to change completely.

metlos committed:
Support for running pre-commit jobs. This is not used yet, but will
be used to optimize the number of required identityhash
re-computations during some bigger transactions that may cause an
identity hash to be recomputed several times unnecessarily atm.

metlos committed:
Redo identity hash computation so that the single algorithm is used.

Before this, the algo was duplicated for MetadataPack.Members and
Entity which was not ideal given the fact that the algorithm is
going to get much more complicated for Merkle tree computation of
many more entity types.

Now a "HashableView" is constructed from either the MT members or
the entity and its dependent entities and this view is then used
to compute the hash.

metlos committed:
Add a shortcut method for unlimited unsorted paging.

metlos committed:
Fix error reporting in some cases during path parsing/extending.

metlos committed:
Make StructuredData.toString() more JSON-like.

metlos committed:
Rework identity-hash computation so that it is usable also outside of
MetadataPack identity computation.

This enables introduction of identity-hashed entities that will be
used for inventory sync.

metlos committed:
Fix Path.equals() and .hashCode() to be immune from start and end
index differences.

metlos committed:
Update headers and make stuff compile. Test failures still present though.

metlos committed:
IdentityHash.treeOf() works now.

metlos committed:
Totally reworked pre-commit actions so that it is possible to only
compose the notifications at the actual commit of the transaction
even in the case of a transaction frame.

This is needed to minimize the number of times we recompute the tree
hash in case of bulk insert, etc.

The transaction is no longer subclassable in backends but instead
they can attach data to it. This then enables the base to have
specialized transaction impls for different transactional behaviors -
like normal and "in-frame" - and not collide with specializations in
backends (which are no longer possible).

The actual re-computation of the notifications based on the results
of tree hashing is NOT implemented yet.

metlos committed:
Update license years.

metlos committed:
Efficient re-computation and update of hashes in pre-commit stage.

metlos committed:
Made all hashable entities hashable.

metlos committed:
Testsuite now passes.

metlos committed:
license updates.

metlos committed:
API checks pass.

metlos committed:
Merge remote-tracking branch 'upstream/master' into bug/HWKINVENT-157

metlos committed:
Fix operation type id deduction in Canonicalpath.ids().

metlos committed:
Fix inspecting data of operation types.

metlos committed:
tmp - sending update and delete events + tests

metlos committed:
Merge branch 'experimental/sql-backend3' into bug/HWKINVENT-157

metlos committed:
Merge remote-tracking branch 'upstream/master' into bug/HWKINVENT-157

metlos committed:
Make stuff compile.

metlos committed:
Don't ignore invalid input during identity hash update.

metlos committed:
Identity hashes recomputed on CUD of children, notifications work.

metlos committed:
Make stuff compile.

metlos committed:
Fix pre-delete of operation types.

metlos committed:
* Fix breadth-first search in tree traversal
* Fix computation of hash changes in presence of deleted entities.

metlos committed:
Disambiguate the REST paths.

metlos committed:
Add isParentOf() to both Canonical and Relative paths.
Added a "from" method to Path.Segment to create one from a string
representation.

metlos committed:
Introduce *.Container interfaces so that we have a common way of
expressing "something that can contain entities of type X".

metlos committed:
Implement the IdentityHashed interface for all the entity types that
should support it.

Add some basic tests for the functionality that should get us going with
it.

metlos committed:
Remove left over traces of relationships in inventory structure.

metlos committed:
Make IdentityHashed.Single extend ResolvableToSingle so that it can be
used in its place on places like Inventory.inspect().

This is to avoid unnecessary unchecked casts that the users would face
if it stayed the way it was.

metlos committed:
Make IdentityHashed interfaces usable with DataEntity, which cannot hold
user-defined relationships.

metlos committed:
Remove the remnants of (non-functional) support for rels in inv structure.

metlos committed:
Support for InventoryStructure JSON serialization.

metlos committed:
Make sure changes to inventory structure are persisted in such a way
that resource types and metric types are persisted prior to resources,
metrics, and other dependent types of entities.

metlos committed:
Implement the REST interface for sync.

metlos committed:
Make checkstyle happy.

metlos committed:
Merge 'master' into bug/HWKINVENT-157

jkremser commented on hawkular-inventory-api/src/test/java/org/hawkular/inventory/api/test/AbstractBaseInventoryTestsuite.java:
```diff
@@ -2181,6 +2470,105 @@ public void testObserveDataEntities() throws Exception {
+                    inventory.observable(Interest.in(OperationType.class).having(identityHashChanged())),
+                    inventory.observable(Interest.in(Resource.class).having(identityHashChanged())),
+                    inventory.observable(Interest.in(Metric.class).having(identityHashChanged())),
+                    inventory.observable(Interest.in(DataEntity.class).having(identityHashChanged()))
```
This is the only place where the `.having(identityHashChanged()))` is used, so it's not turned on by default?

metlos replied:
Whoops, good catch.. Bus integration should have that...

metlos replied:
Should be there by commit 4e04f20.

jkremser commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/TreeTraversal.java:
```diff
@@ -0,0 +1,118 @@
+        traverse(node, queue, visitor);
+    }
+
+    private void traverse(Node node, Memory<Iterator<Node>> memory, Function<Node, Boolean> visitor) {
```
perhaps this can be even public so that user can pass some own impl of the "memory" ..

(btw. I don't like the name "memory", perhaps something like "DataStructure"  would be more suitable, if it stays as an impl. details I am ok with `Memory`, but if the method is public, I'd rather rename it)

..like priority queue

metlos replied:
Well, I called it `Memory` because it's used to remember positions in the tree traversal. Maybe we could call it `BackLog` or even maybe `Order`. I don't like `DataStructure` though because that doesn't tell anything about the purpose of that class.

jkremser commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/IdentityHash.java:
```diff
@@ -0,0 +1,810 @@
+import org.hawkular.inventory.paths.SegmentType;
+
+/**
+ * Produces an identity hash of entities. Identity hash is a hash that uniquely identifies an entity
```
~~`uniquely identifies an entity` Does it mean, there can't be a hash collision?~~

EDIT: I got my answer below :) It's SHA1 of the string concatenation of entity's significant fields + hashes of the structure (successors). So there could be a collision, but it's unlikely. 

What about:

`hash that with high probability uniquely identifies an entity`?

btw. http://stackoverflow.com/a/1867252/1594980

so it means for 1M of entities, the probability of collision is
https://www.wolframalpha.com/input/?i=n(n-1)%2F2+*+1%2F2%5Eb+where+n+%3D+1000000,+b+%3D+160
.. pretty low :]

metlos replied:
Well, it's the same thing git uses to identify commits, so I'd argue we're pretty safe here. Probability that in a single tenant there will be two different entities of the same type with the same identity hash is pretty low ;)

jkremser commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/IdentityHash.java:
```diff
@@ -0,0 +1,810 @@
+//            System.out.println("start: " + context.root);
+        }
+
+        public void endChild(IntermediateHashContext ctx, IntermediateHashResult result) {
```
shouldn't the class be abstract and methods `startChild` and `endChild` too?

metlos replied:
I left the commented out debug statements there because I had so much trouble getting the computation right that I thought someone might use that again in the future ;)

This is a private class so I was not that fussed about purity but you're right - this could be abstract. It could even be made a member class of the `treeOf` method because that seems to be the only usage of it.

metlos committed:
Merge branch 'master' into bug/HWKINVENT-157

jkremser commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/InventoryStructure.java:
```diff
@@ -0,0 +1,847 @@
+                        return ElementTypeVisitor.accept(parentType, new ElementTypeVisitor.Simple<Stream<BB>, Void>() {
+                            @Override public Stream<BB> visitEnvironment(Void parameter) {
+                                return fromRead(Environment.class, Environments.Single.class,
+                                        Metrics.Container::metrics);
```
this is insane, sick, sick :)

metlos committed:
[HWKINVENT-157] Publish identityHashChanged events on the bus.

jkremser commented on hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/InventoryStructure.java:
```diff
@@ -0,0 +1,847 @@
+     * This is not directly instantiable but rather can be either copied from another structure (possibly lazily
+     * loaded) or built using a {@link Builder}.
+     */
+    class Offline<Root extends Entity.Blueprint> implements InventoryStructure<Root>, Serializable {
```
Does it mean, we can do some kind of snapshots of the inventory structure and save/load it.

metlos replied:
Technically, yes. But this is not the full inventory, just the stuff that is identity-hashable (which is everything under a feed, but tenant and environment aren't hashable (now)).

jkremser commented on hawkular-inventory-rest-api/src/main/java/org/hawkular/inventory/rest/RestSync.java:
```diff
@@ -0,0 +1,71 @@
+                    + " are not synchronizable.");
+        }
+
+        inventory.inspect(cp, IdentityHashed.SingleWithRelationships.class).synchronize(structure);
```
Does it take long to do this? Perhaps, in the future we may do the `/sync` endpoint async w/ the `@Suspended AsyncResponse response`

metlos replied:
It very much can take a long time.

jkremser commented on hawkular-inventory-json-helper/src/main/java/org/hawkular/inventory/json/InventoryStructureDeserializer.java:
```diff
@@ -0,0 +1,129 @@
+                    continue;
+                }
+
+                InventoryStructure.EntityType type = InventoryStructure.EntityType.valueOf(typeName);
```
shouldn't this be also put into `try-catch` block? Like on line 70?

``` java
        InventoryStructure.EntityType type;
        try {
            type = InventoryStructure.EntityType.valueOf(typeName);
        } catch (IllegalArgumentException e) {
            throw new JsonParseException("Unrecognized value of 'type'. Supported values are " + LEGAL_ENTITY_TYPES
                     + " but got '" + typeName + "'.", JsonLocation.NA);
         }
```

metlos replied:
done in 1d8fd51ae609b93bb8567e335e6786bb2981eea6.

jkremser commented on hawkular-inventory-itest/src/test/groovy/org/hawkular/inventory/rest/test/InventoryITest.groovy:
```diff
@@ -1291,6 +1292,90 @@ class InventoryITest extends AbstractTestBase {
+                                                id                : "metricType",
+                                                type              : "GAUGE",
+                                                unit              : "NONE",
+                                                collectionInterval: 0,
```
offtopic: perhaps we should rename the `collectionInterval` on metricType entity to `defaultCollectionInterval`

@pavolloffay what do you think? Does it make sense?

metlos replied:
If we did that, I'd vote for renaming `metricType.type` to `metricType.kind`.

jkremser commented on hawkular-inventory-itest/src/test/groovy/org/hawkular/inventory/rest/test/InventoryITest.groovy:
```diff
@@ -1291,6 +1292,90 @@ class InventoryITest extends AbstractTestBase {
+            def response = client.post(path: "$basePath/feeds", body: "{\"id\": \"sync-feed\"}")
+            assertEquals(201, response.status)
+
+            response = client.post(path: "$basePath/sync/f;sync-feed", body: structure)
```
+1 the REST api looks pretty easy to use. And for creating the JSON there is the builder in place.

metlos replied:
Yeah, I went ahead and tried what it'd look like if we implemented #201. How is your schedule looking with regards to that? I'd very much like to go forward with #201. If you think you won't find time for it I'm happy to take over...

jkremser replied:
I am snowed under w/ Ruby, if you don't mind, go ahead

metlos committed:
[HWKINVENT-157] throw a JsonParseException instead of just IAE
when encountering invalid child data type during inventory structure
deserialization.

jkremser commented:
```
/home/jkremser/workspace/hawkular/hawkular-inventory/hawkular-inventory-json-helper/src/main/java/org/hawkular/inventory/json/InventoryStructureDeserializer.java:110: Line is longer than 120 characters (found 123).
```

metlos committed:
Checkstyle happy again.

jkremser commented:
``` bash
🏃 m clean install -Pitest
...
[INFO] Reactor Summary:
[INFO] 
[INFO] hawkular-inventory-parent .......................... SUCCESS [  5.341 s]
[INFO] Hawkular Inventory Paths ........................... SUCCESS [  2.731 s]
[INFO] Hawkular Inventory API ............................. SUCCESS [  9.985 s]
[INFO] Hawkular Inventory JSON Helper ..................... SUCCESS [  2.727 s]
[INFO] Hawkular Inventory Bus API ......................... SUCCESS [  1.227 s]
[INFO] Hawkular Inventory CDI Integration ................. SUCCESS [  1.776 s]
[INFO] Hawkular Inventory Tinkerpop implementation parent POM SUCCESS [  0.723 s]
[INFO] Hawkular Inventory Tinkerpop SPI ................... SUCCESS [  0.964 s]
[INFO] Hawkular Inventory Tinkerpop Implementation ........ SUCCESS [  1.830 s]
[INFO] Hawkular Inventory Tinkerpop TinkerGraph Provider .. SUCCESS [  8.681 s]
[INFO] Hawkular Inventory Bus Integration ................. SUCCESS [  2.829 s]
[INFO] Hawkular Inventory REST API Web Application ........ SUCCESS [  3.985 s]
[INFO] Hawkular Inventory Tinkerpop Titan Provider ........ SUCCESS [07:19 min]
[INFO] Hawkular Inventory REST War Distribution ........... SUCCESS [  3.942 s]
[INFO] hawkular-inventory-feature-pack .................... SUCCESS [  1.497 s]
[INFO] Hawkular Inventory Tinkerpop SQL Provider .......... SUCCESS [ 29.116 s]
[INFO] hawkular-inventory-itest ........................... SUCCESS [01:05 min]
[INFO] Hawkular Inventory Load Tests ...................... SUCCESS [  1.866 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 09:45 min
[INFO] Finished at: 2016-04-13T02:45:18+02:00
[INFO] Final Memory: 196M/526M
[INFO] ------------------------------------------------------------------------
```

It's passing for me locally, so I am merging..
